src/client/app_configuration_ibm_cloud.rs

```diff
@@ -229,35 +229,15 @@ impl AppConfigurationClient for AppConfigurationClientIBMCloud {
         let feature = config_snapshot.get_feature(feature_id)?;
 
         // Get the segment rules that apply to this feature
-        let segments = {
-            let all_segment_ids = feature
-                .segment_rules
-                .iter()
-                .flat_map(|targeting_rule| {
-                    targeting_rule
-                        .rules
-                        .iter()
-                        .flat_map(|segment| &segment.segments)
-                })
-                .cloned()
-                .collect::<HashSet<String>>();
-            let segments: HashMap<String, Segment> = config_snapshot
-                .segments
-                .iter()
-                .filter(|&(key, _)| all_segment_ids.contains(key))
-                .map(|(k, v)| (k.clone(), v.clone()))
-                .collect();
-
-            // Integrity DB check: all segment_ids should be available in the snapshot
-            if all_segment_ids.len() != segments.len() {
-                return Err(ConfigurationAccessError::MissingSegments {
-                    resource_id: feature_id.to_string(),
-                }
-                .into());
-            }
+        let segments = config_snapshot.get_segments_for_segment_rules(&feature.segment_rules);
 
-            segments
-        };
+        // Integrity DB check: all segment_ids should be available in the snapshot
+        if feature.segment_rules.len() != segments.len() {
+            return Err(ConfigurationAccessError::MissingSegments {
+                resource_id: feature_id.to_string(),
+            }
+            .into());
+        }
```

Comment:
can we have segment rules referencing the same segments?
In this case this would fail.